### PR TITLE
Support creating proofs that can be verified on Ethereum

### PR DESCRIPTION
**Pull request by mellowcroc**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,18 +43,59 @@ Currently, only `linux/amd64` with `AVX` is supported.
...
 
 Using `--network starknet` serializes the Cairo proof into a format that can be verified on the Cairo verifier deployed on Starknet. Please refer to the [integrity documentation](https://github.com/HerodotusDev/integrity) for more information on how to use the calldata to send a transaction to Starknet.
 
+Using `--network ethereum` serializes the Cairo proof into a format that can be verified on the Cairo verifier deployed on Ethereum. Please refer to the [the next section](#how-to-create-proofs-and-verify-them-on-ethereum) for more information on how to create proofs that can be verified on Ethereum.
+
+### How to create proofs and verify them on Ethereum
+
+Currently there is a Cairo verifier deployed on Ethereum, which are mainly used to verify SHARP proofs created by L2 Starknet nodes. The Cairo verifier checks the validity of a Cairo program named `bootloader`, which recursively verifies multiple Cairo programs or a Cairo PIEs (Position Independent Executable) and allows a single proof to prove executions of multiple Cairo programs. Once we create a bootloader proof, we need to serialize it to a format that works for the Cairo verifier on Ethereum.
```
Comment (rot256) [inline README.md:79]: "recursively verifies". 

My understanding is that that the bootloader just hashes the program and jumps to it?

File: src/bootloader.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        relocate_mem: true,
+        layout,
+        proof_mode: true,
+        secure_run: None,
+        disable_trace_padding: false,
+        allow_missing_builtins: None,
+    };
+
+    let program_hash = Felt252::from_dec_str(
+        "382450030162484995497251732956824096484321811411123989415157331925872358847",
```
Comment (rot256) [inline src/bootloader.rs:168]: Move into a constant, do we have a public reference for this constant we can include in a comment?

File: src/bootloader.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    )
+    .unwrap();
+    let verifier_hashes_hex = vec![
```
Comment (rot256) [inline src/bootloader.rs:171]: Move into a constant, do we have a public reference for this constant we can include in a comment?

File: src/prover.rs [hunk 4]
```diff
@@ -28,28 +28,86 @@ pub fn run_stone_prover(
...
+    tmp_dir: &tempfile::TempDir,
+) -> Result<(), ProverError> {
     let tmp_prover_parameters_path = tmp_dir.path().join("prover_parameters.json");
 
-    let prover_parameters_path = if let Some(parameter_file) = &prove_args.parameter_file {
+    let prover_parameters_path = if let Some(parameter_file) = &parameter_file {
         parameter_file
     } else {
         let air_public_input_json: PublicInput =
             serde_json::from_str(&fs::read_to_string(air_public_input)?).unwrap();
```
Comment (rot256) [inline src/prover.rs:98]: The PR did not touch this, but seems to panic if the file is not JSON?

File: src/serialize.rs [hunk 5]
```diff
@@ -1,43 +1,104 @@
...
+    #[error("Annotation file is required for serializing proofs for Ethereum")]
+    AnnotationFileNotSpecified,
+    #[error("Extra output file is required for serializing proofs for Ethereum")]
+    ExtraOutputFileNotSpecified,
+}
+
+pub fn serialize_proof(args: SerializeArgs) -> Result<(), Error> {
     let proof_file = args.proof.clone();
-    let (config, public_input, unsent_commitment, witness) = parse_proof_file(&proof_file)?;
+    if args.network == Network::ethereum {
```
Comment (rot256) [inline src/serialize.rs:31]: I would use a `match` here as you are trying to cover every variant.

File: src/utils.rs [hunk 6]
```diff
@@ -96,3 +95,34 @@ impl FileWriter {
         self.buf_writer.flush()
     }
 }
+
+// `CairoRunner.get_air_public_input()` returns a `PublicInput` object.
+// This function converts it to a JSON string and formats the "public_memory" array
+// by prefixing each value with "0x" if it doesn't already start with "0x".
+pub fn get_formatted_air_public_input(air_public_input: &str) -> Result<String, Error> {
```
Comment (rot256) [inline src/utils.rs:102]: Could this function take a `PublicInput` struct instead?

**Review by rot256 — APPROVED** [review]
This is really nice!

I only had a few more/less pedantic comments.

**Comment by rot256** [comment]
Is this PR stale or are we good to merge?

**Comment by mellowcroc** [comment]
Good to merge.

**Pull request merged**